/* Copyright 2022 The MediaPipe Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

syntax = "proto2";

// TODO: consolidate TensorToSegmentationCalculator.
package mediapipe.tasks;

import "mediapipe/framework/calculator_options.proto";
import "mediapipe/tasks/cc/vision/image_segmenter/proto/segmenter_options.proto";
import "mediapipe/util/label_map.proto";

option java_package = "com.google.mediapipe.tasks";
option java_outer_classname = "TensorsToSegmentationCalculatorOptionsProto";

message TensorsToSegmentationCalculatorOptions {
  extend mediapipe.CalculatorOptions {
    optional TensorsToSegmentationCalculatorOptions ext = 458105876;
  }

  optional mediapipe.tasks.vision.image_segmenter.proto.SegmenterOptions
      segmenter_options = 1;

  // Identifying information for each classification label.
  map<int64, mediapipe.LabelMapItem> label_items = 2;

  // Options used for confidence masks only. Ignored for category masks.
  //
  // Example: Assume there are 6 input channels.
  // By default, 6 output streams are generated and input channel n goes to
  // output stream n.
  // For
  //   confidence_mask_options {
  //     output_channels: 2
  //     output_channels: 3
  //     output_channels: 4
  //     pack4: true
  //     output_format: OUTPUT_FORMAT_UINT8
  //   }
  // only one uint8 RGBA output stream is generated, with channels 2, 3 and 4
  // packed into the RGB channels of the output stream. The A channel is left
  // blank (zero).
  message ConfidenceMaskOptions {
    // Specifies which channels to send to the output. The number of
    // output_channels determines the number of output streams. Channels that
    // are not specified will be discarded (but used for intermediate math like
    // SOFTMAX). If unset, all input channels are sent to the output.
    repeated int32 output_channels = 1 [packed = true];

    // If false (default), outputs one mask for each channel. If true, packs 4
    // consecutive output channels into a single image.
    optional bool pack4 = 2;

    enum OutputFormat {
      OUTPUT_FORMAT_UNSPECIFIED = 0;  // Default is FLOAT.
      OUTPUT_FORMAT_FLOAT = 1;  // Float32 or Half16, depending on GPU support.
      OUTPUT_FORMAT_UINT8 = 2;  // Uint8.
    }

    // Specifies the format of the output images.
    // Note that OUTPUT_FORMAT_UINT8 with pack4==false (single channel 8 bit)
    // output with GPU processing may not be supported on all platforms.
    optional OutputFormat output_format = 3;
  }

  optional ConfidenceMaskOptions confidence_mask_options = 3;
}
